text/0000-pin_and_move.md

```diff
@@ -0,0 +1,381 @@
+
+### Notes on existing ecosystem traits
+
+These traits supplant certain traits in the ecosystem which already provide
+similar guarantees. In particular, the [stable_deref_trait][stable-deref]
+crate provides a similar bit different hierarchy. The differences are:
+
+- That crate draws no distinction between `StableDeref` and `StableDerefMut`.
+  This does not leave forward compatibility for the static string optimization
+  mentioned previously.
+- That crate has no equivalent to the `Own` trait, which is necessary for some
+  APIs using internal references.
+- That crate has a `StableDerefClone` type, which is equivalent to the bound
+  `Share + StableDeref` in our system.
+
+If the hierarchy proposed in this RFC becomes stable, all users are encouraged
+to migrate from that crate to the standard library traits.
+
+## The `Pin` type
+
+The `Pin` type is a wrapper around a mutable reference. If the type it
+references is `!Move`, the `Pin` type guarantees that the referenced data will
+never be moved again. It has a relatively small API. It is added to both
+`std::mem` and `core::mem`.
+
+```rust
+struct Pin<'a, T: ?Sized + 'a> {
+    data: &'a mut T,
+}
+
+impl<'a, T: ?Sized + Move> Pin<'a, T> {
+    pub fn new(data: &'a mut T) -> Pin<'a, T>;
+}
+
+impl<'a, T: ?Sized> Pin<'a, T> {
+    pub unsafe fn new_unchecked(data: &'a mut T) -> Pin<'a, T>;
+
+    pub unsafe fn get_mut(this: Pin<'a, T>) -> &'a mut T;
+
+    pub fn borrow<'b>(this: &'b mut Pin<'a, T>) -> Pin<'b, T>;
+}
+
+impl<'a, T: ?Sized> Deref for Pin<'a, T> {
+    type Target = T;
+}
+
+impl<'a, T: ?Sized + Move> DerefMut for Pin<'a, T> { }
+```
+
+For types which implement `Move`, `Pin` is essentially the same as an `&mut T`.
+But for types which do not, the conversion between `&mut T` and `Pin` is
+unsafe (however, `Pin` can be easily immutably dereferenced, even for `!Move`
+types).
+
+The contract on the unsafe part of `Pin`s API is that a Pin cannot be
+constructed if the data it references would ever move again, and that it cannot
+be converted into a mutable reference if the data might ever be moved out of
+that reference. In other words, if you have a `Pin` containing data which does
+not implement `Move`, you have a guarantee that that data will never move.
+
+The next two subsections describe safe APIs for constructing a `Pin` of data
+which cannot be moved - one in the heap, and one in the stack.
+
+### Pinning to the heap: The `Anchor` type
+
+The `Anchor` wrapper takes a type that implements `StableDeref` and `Own`, and
+prevents users from moving data out of that unless it implements `Move`. It is
+added to `std::mem` and `core::mem`.
+
+```rust
+struct Anchor<T> {
+     ptr: T,
+}
+
+impl<T: StableDerefMut + Own> Anchor<T> {
+     pub fn new(ptr: T) -> Anchor<T>;
+
+     pub unsafe fn get_mut(this: &mut Anchor<T>) -> &mut T;
+
+     pub unsafe fn into_inner_unchecked(this: Anchor<T>) -> T;
+
+     pub fn pin<'a>(this: &'a mut Anchor<T>) -> Pin<'a, T::Target>;
```

Comment:
Did you consider making this a method, rather than an associated function taking `this`? We generally avoid adding methods to `Deref` types, but in this case, it seems like `pin` would be called more than any other function on an `Anchor` type (and if you do call `pin`, you can then call any `&self` method, as `Pin` is also `Deref`).

Reply:
Also, `pin` seems like a misleading name for this function, since it isn't actually pinning the data anywhere. `Anchor::new` pins the data to the heap-- `pin` just gives you a reference to the object as a `Pin`. Perhaps `as_pin` would be better?

Reply:
I have no commitment on either of these questions!